From 78f8ad8cda48a4dd279e682f49a66f0c63908d5c Mon Sep 17 00:00:00 2001
From: Biju Chakkachattil <bijux.chakkachattil@intel.com>
Date: Mon, 22 Jun 2015 04:27:20 +0530
Subject: [PATCH] EM/PMIC: Check USB Host cable state during boot.

While booting with OTG connected, it is not detected because the
initial events were missed during booting. This patch will check
the USB host cable state during boot and schedule the work if OTG
is connected.

Change-Id: Ic608a28d4b5b0b0716ce1b49dd41318fb57897fa
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-36254
Signed-off-by: Biju Chakkachattil <bijux.chakkachattil@intel.com>
---
 drivers/power/intel_pmic_ccsm.c | 21 ++++++++++++++++-----
 1 file changed, 16 insertions(+), 5 deletions(-)

diff --git a/drivers/power/intel_pmic_ccsm.c b/drivers/power/intel_pmic_ccsm.c
index 5339622c4536..4b8dcd45dba5 100644
--- a/drivers/power/intel_pmic_ccsm.c
+++ b/drivers/power/intel_pmic_ccsm.c
@@ -1531,6 +1531,7 @@ static int pmic_check_initial_events(void)
 	u8 val, sreg_val = 0;
 	u16 *pmic_int, *pmic_int_stat, off;
 	u16 stat_reg = 0;
+	struct extcon_dev *edev;
 
 	evt = kzalloc(sizeof(struct pmic_event), GFP_KERNEL);
 	if (!evt)
@@ -1566,8 +1567,18 @@ static int pmic_check_initial_events(void)
 
 	INIT_LIST_HEAD(&evt->node);
 	list_add_tail(&evt->node, &chc.evt_queue);
-	schedule_delayed_work(&chc.evt_work, 0);
 
+	edev = extcon_get_extcon_dev("usb-typec");
+
+	if (!edev)
+		dev_err(chc.dev, "No edev found");
+	else {
+		chc.cable_state = extcon_get_cable_state(edev, "USB-Host");
+		if (chc.cable_state)
+			schedule_work(&chc.extcon_work);
+	}
+
+	schedule_delayed_work(&chc.evt_work, 0);
 	pmic_bat_zone_changed();
 
 	return ret;
@@ -1800,15 +1811,15 @@ static int pmic_chrgr_probe(struct platform_device *pdev)
 	INIT_DELAYED_WORK(&chc.evt_work, pmic_event_worker);
 	INIT_LIST_HEAD(&chc.evt_queue);
 
-	ret = pmic_check_initial_events();
-	if (ret)
-		goto otg_req_failed;
-
 	INIT_WORK(&chc.extcon_work, pmic_ccsm_extcon_host_work);
 	chc.cable_nb.notifier_call = pmic_ccsm_usb_host_nb;
 	extcon_register_interest(&chc.host_cable, "usb-typec", "USB-Host",
 						&chc.cable_nb);
 
+	ret = pmic_check_initial_events();
+	if (ret)
+		goto otg_req_failed;
+
 	/* register interrupt */
 	for (i = 0; i < chc.irq_cnt; ++i) {
 		ret = request_threaded_irq(chc.irq[i], pmic_isr,
-- 
1.9.1

